### ISSUE TITLE
[v2] Unable to intercept service

### ISSUE THREAD

**Issue by andrei-dascalu** [telepresenceio/telepresence#1721]
Hello,

I'm having a really strange problem with t2 in a cluster in that even though I am intercepting a service and t2 says it's all good, requests going to the service don't reach my local environment and still go to the actual service in the cluster.

How it goes:

```
telepresence connect
```
-> connects to traffic manager, successfully using my current kubectl context
-> I am able to `curl` service in the cluster (eg: `curl -XGET http://my-service.my-ns.svc.cluster.local`)

```
telepresence intercept my-service -n my-ns --port 8080:80
```

```
Using Deployment my-service
intercepted
    Intercept name         : my-service-my-ns
    State                  : ACTIVE
    Workload kind          : Deployment
    Destination            : 127.0.0.1:8080
    Service Port Identifier: 80
    Volume Mount Error     : macFUSE 4.0.5 or higher is required on your local machine
    Intercepting           : all TCP connections
```

Note: I have macFUSE 4.1 installed locally, yet telepresence always complains about it.

Now however when the my-service is supposed to receive requests (other services call it as `http://my-service.my-ns.svc` since it exposes port 80 in the cluster), I would expect that my local app which runs on 127.0.0.1:8080 would receive the requests, yet they still go to the cluster service.

The service responds to internal requests, it's not directly involved with ingresses but I can force requests to happen. 

Note: if I create an ad-hoc ingress to expose it via Ambassador Cloud, then requests going to that endpoint do reach my local service.

It's quite strange (particularly as I was able to follow the tutorial with the sample applications on the same cluster).

**Comment by thallgren** [comment]
Do you get the same error if you use a symbolic port, e.g. `targetPort: http` in the service and `name: http` on the container's `ports/containerPort`?

**Comment by andrei-dascalu** [comment]
This is my service spec:

```yaml
spec:
  ports:
    - name: web
      protocol: TCP
      port: 80
      targetPort: 8080
```

and the port configuration as per deployment (which reflects in the pods themselves)

```yaml
      containers:
        - name: my-service-container
          image: '<image:label>'
          ports:
            - containerPort: 8080
              protocol: TCP
```


this is the traffic agent that was added under the current situation

```yaml
        - name: traffic-agent
          image: 'docker.io/datawire/tel2:2.2.1'
          args:
            - agent
          ports:
            - name: tx-8080
              containerPort: 9900
              protocol: TCP
```

I'm not using a symbolic name to target the port, just the exact values, but I can try using one. It would be strange to make a different, wouldn't it ?

**Comment by thallgren** [comment]
Regarding macFUSE, yes, we're aware of it: #1651 and will hopefully have a fix for it soon.

**Comment by thallgren** [comment]
Your setup should work using port numbers only. I just wanted to narrow down the cause of the problem.

**Comment by andrei-dascalu** [comment]
I've switched to using a symbolic name for the port, but seems it doesn't make a difference

**Comment by thallgren** [comment]
Ok. Thanks for trying that.

**Comment by andrei-dascalu** [comment]
I have some additional info, though no explanation for what I saw.

Even after doing in order:
- telepresence leave <my intercept>  -> which has no output to confirm success
- telepresence quit

I had two deployments that were left with telepresence managedFields entry and with sidecar. The kicker is that one of them has values for telepresence 2.2.0 and the other for 2.2.1 (the second one is the one for which I opened this ticket). During the 2.2.0 to 2.2.1 upgrade I initially did an `uninstall --everything` and now I did it again,  but I still have the fields for both the services (now I am technically uninstalling 2.2.1 but the uninstall is leaving the fields in my deployment).

**Comment by thallgren** [comment]
That sounds like a bug for sure. Would it be possible for you to convey that deployment manifest?

**Comment by andrei-dascalu** [comment]
uhm, in the meantime I performed a manual cleanup, then reinstalled everything after looking not to have leftovers traffic managers or annotations.

I did everything from scratch, including the targeted deployment (except the cluster itself) but I still have the issue.

Now I wanted to uninstall again, but I have:

```
❯ telepresence uninstall --everything                                                                                                                                                                                        
telepresence: error: Deployment name="my-service" namespace="my-ns": unable to locate port "tm-appport" in container "my-service-container"
```

```yaml
      containers:
        - name: my-service-container
          image: '<image:label>'
          ports:
            - name: appport
              containerPort: 8080
              protocol: TCP
```

However, the agent has no port "tm-appport", what it has is

```
          ports:
            - name: appport
              containerPort: 9900
              protocol: TCP
```

**Comment by andrei-dascalu** [comment]
Yaml as it seems to have been left hanging by 2.2.0
I edited out some potentially sensitive information. Seems just the volume and the agent was left by the previous version and 2.2.1 picks it up as already installed (though missing annotations and managed fields - it has just volumes and the agent sidecar) and won't remove it or upgrade it.

```yaml
kind: Deployment
apiVersion: apps/v1
metadata:
  name: dev-my
  namespace: my-ns
  generation: 11
  creationTimestamp: '2021-04-20T10:04:49Z'
  labels:
    app: backend
    app.kubernetes.io/instance: my-dev-instance
    env: dev
  annotations:
    deployment.kubernetes.io/revision: '7'
    ... no annotation for telepresence actions
  managedFields:
    ... no managed fields for telepresence
spec:
  replicas: 1
  selector:
    matchLabels:
      app: backend
      env: dev
  template:
    metadata:
      labels:
        app: backend
        env: dev
    spec:
      volumes:
        - name: traffic-annotations
          downwardAPI:
            itmy:
              - path: annotations
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.annotations
            defaultMode: 420
      containers:
        - name: apache-exporter
          image: 'bitnami/apache-exporter:0.8.0'
          command:
            - /bin/bash
            - '-c'
            - 'apache_exporter --scrape_uri="http://127.0.0.1/server-status?auto"'
          ports:
            - name: metrics-port
              containerPort: 9117
              protocol: TCP
          resources:
            limits:
              cpu: 100m
              memory: 64Mi
            requests:
              cpu: 50m
              memory: 64Mi
          imagePullPolicy: Always
        - name: my-app
          image: '<mycontainer>'
          ports:
            - name: web-port
              containerPort: 80
              protocol: TCP
          envFrom:
            - configMapRef:
                name: config-common
            - secretRef:
                name: secret-common
            - configMapRef:
                name: config-dev
            - secretRef:
                name: secret-dev
          resources:
            limits:
              cpu: 500m
              memory: 512Mi
            requests:
              cpu: 250m
              memory: 256Mi
          lifecycle:
            postStart:
              exec:
                command:
                  - /bin/bash
                  - '-c'
                  - cd /var/www/html && /usr/bin/post-start
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
        - name: traffic-agent
          image: 'docker.io/datawire/tel2:2.2.0'
          args:
            - agent
          ports:
            - name: web-port
              containerPort: 9900
              protocol: TCP
          envFrom:
            - prefix: TEL_APP_
              configMapRef:
                name: config-common
            - prefix: TEL_APP_
              secretRef:
                name: secret-common
            - prefix: TEL_APP_
              configMapRef:
                name: config-dev
            - prefix: TEL_APP_
              secretRef:
                name: secret-dev
          env:
            - name: TELEPRESENCE_CONTAINER
              value: my-app
            - name: LOG_LEVEL
              value: debug
            - name: AGENT_NAME
              value: dev-my
            - name: AGENT_POD_NAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.name
            - name: AGENT_NAMESPACE
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
            - name: APP_PORT
              value: '80'
            - name: MANAGER_HOST
              value: traffic-manager.ambassador
          resources: {}
          volumeMounts:
            - name: traffic-annotations
              mountPath: /tel_pod_info
          readinessProbe:
            exec:
              command:
                - /bin/stat
                - /tmp/agent/ready
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 3
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      securityContext: {}
      schedulerName: default-scheduler
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 0
      maxSurge: 1
  revisionHistoryLimit: 3
  progressDeadlineSeconds: 600
```

**Comment by andrei-dascalu** [comment]
Ok, so I got it working!!!

I'm going to leave a few observations, not sure if any/all/none may be linked to why it didn't work - or whether they are problematic overall.

1. The deployment stuck with a 2.2.0 sidecar isn't the one I was trying to work with, it would be weird if this was the cause
2. My attempts to manually cleanup this deployment were undermined by the fact that I was restarting telepresence daemon BEFORE updated (clean) pods were running. I guess T2 looked at the running pods directly and saw they had an agent and left it alone (though it was missing annotations and whatnot)
3. Subsequent "uninstall" ran via T2 2.2.1 CLI tool ignored this deployment and left the sidecar there (possibly due to missing annotations ?)
4. My actual target service seemed to act badly when I used a named port to expose (not when I used named port on deployment/pods or in the service to target the pods). The T2 sidecar seemed stuck on that even when I started using only port numbers across the board (`unable to find target port "web.tx-8080"` - where web was the name of the port as exposed by the service, even when I removed it)
5. When I was doing `telepresence leave <intercept>` and then tinker with ports, changes were ignored (the above point is a symptom of this) when I did intercept again .... I guess I would need to quit the daemon? Not sure if that would work.

Bottom line ... it started working when I dumped named ports and used numbers everywhere.

**Comment by andrei-dascalu** [comment]
Now when I try to `uninstall --everything`, I keep getting

`telepresence: error: Service name="my-service" namespace="my-ns": unable to find target port "tx-8080"`

For no apparent reason, after I was successfully working with it
Did  `leave` then `uninstall --everything` and got this error. Now it's back to not working at all when I try to intercept.

**Comment by andrei-dascalu** [comment]
The usage pattern I now see is this:

- freshly setup T2 and traffic-manager in cluster
- intercept service
- play around, it works
- stop the local service, start it again - usually `leave` and `intercept` again work, but after a couple of time it just doesn't work, has no more effect
- `uninstall --everything` shows the error about `unable to find port` even though the port seems to exist

```yaml
        - name: traffic-agent
          image: 'docker.io/datawire/tel2:2.2.1'
          args:
            - agent
          ports:
            - name: tx-8080
              containerPort: 9900
              protocol: TCP
```

solution:
- stop T2 local daemon
- manually delete `ambassador` namespace (or svc/deployment as per uninstall instructions)
- delete affected deployment and redeploy it
- start T2 again, it all works again.

**Comment by donnyyung** [comment]
Hey @andrei-dascalu, I see that the port is in your workload (deployment, replicaSet, etc.) but the error about `telepresence: error: Service name="my-service" namespace="my-ns": unable to find target port "tx-8080` is saying it can't find "tx-8080" in the service.  Is there a chance you (or something) applied a change to the service which overwrote the edits Telepresence makes?

**Comment by andrei-dascalu** [comment]
Hi @donnyyung I thought the same but it wasn't the case. At the time of the error I manually checked both deployment and pods themselves but they all had all the things Telepresence adds (volume/sidecar with the expected ports/annotations/etc).

I also had a peer debugging session with Peter and double checked that the last change since an intercept is the one that T2 does.

The pods did not restart since intercept was started.

**Comment by peteroneilljr** [comment]
Hi @donnyyung I looked at this with @andrei-dascalu the behavior and workaround look very similar to https://github.com/telepresenceio/telepresence/issues/1698 but the steps to get there are different. I would say the underlying problem is likely related.

**Comment by andrei-dascalu** [comment]
A couple more observations, not sure if relevant or not, but I've tried a similar setup in a different cluster with somewhat different results (at least statistically).

Cluster 1, where the reported issue happens:
- Azure, k8s 1.19
- using flux (but flux disabled for the service that was intercepted)
- frequency of issue: 100% (could happen as soon as 5 minutes after intercept or as late as ~ 10 min, intercept works as expected for just a bit, but then stops without visible errors in logs even though the deployment & pods seem unaltered)

Cluster 2, tried today:
- GKE, k8s 1.19
- no flux at all
- frequency of issue: ~ 25% - tried about 20 intercepts in a similar setup, eventually 5 of them failed silently with the same symptoms (nothing in logs, error about missing port though port was there, serviced pods did not restart since sidecar was added) but overall the intercepts worked for significantly longer.

**Comment by wasd171** [comment]
Having a similar issue on Azure k8s 1.19.7

**Comment by alan994** [comment]
Any update on this? I think I have this issue too.

@andrei-dascalu @peteroneilljr @donnyyung @thallgren

**Comment by luminoso** [comment]
I don't know if it makes sense, but for me its solved since 2.7.6 if I recall correctly

**Comment by Avi1235** [comment]
Having the same issue, where can i download 2.7.6 version?

**Comment by thallgren** [comment]
@Avi1235 the latest release of Telepresence is 2.4.10.

**Comment by thallgren** [comment]
@alan994 the only way I can see these port problems happening is if you edit the pod template or service manifest manually.

When telepresence intercepts a service where the service's `targetPort` is symbolic, it will:
1. Rename the container's `port` matching that port to "tm-xxx" (where "xxx") is the original name.
2. Install an agent with a port named "xxx".
3. Add annotations that describe what was done so that it can be undone.

If the `targetPort` is numeric, then telepresence needs to:
1. change the `targetPort` to "tm-nnn" (where "nnn" is the original port number), i.e. it becomes a symbolic port.
2. Install an agent with a port named "tm-nnn".
3. Add annotations that describe what was done so that it can be undone.

In both cases this means that the service now forwards to the agent instead of the container. The agent will then forward to the container unless an intercept is active. It also means that if the workload is edited manually, then the uninstall will break just like described in this bug.

**Comment by alan994** [comment]
@thallgren you are probably right. In our case, we used [dapr](https://dapr.io/). Dapr also modifies `deployment` and I think that was the cause of my problem. We decided to ditch `dapr` and make some changes to our cluster. When this process finishes, I will try again and let you know if my problem is solved. Thanks again!

**Comment by cindymullins-dw** [comment]
It appears something was changing the pod template, perhaps DAPR.  Using symbolic ports is recommended in general.  Closing this issue, please reopen the issue persist on newer versions.
